src/extension/neoExpress/neoExpressInstaller.ts

```diff
@@ -0,0 +1,181 @@
+    let currentPackage = await findPackage(this.targetPackage, VersionMatchCriteria.nameOnly);
+    if (currentPackage) {
+      const updateOutput = await this.tryUpdate(currentPackage, {
+        ...this.targetPackage,
+        location: currentPackage.location,
+      });
+      if (updateOutput.package) {
+        currentPackage = updateOutput.package;
+      }
+    } else {
+      currentPackage = await this.tryInstall();
+    }
+    return currentPackage;
+  }
+
+  async tryInstall(): Promise<DotNetPackage | null> {
+    let installedPackage = null;
+    try {
+      const selectedLocation = await this.selectUserPreferredLocation();
+      if (selectedLocation === null) {
+        return null;
+      }
+      installedPackage = { ...this.targetPackage, location: selectedLocation };
+      await installCommand(this.rootFolder, installedPackage, getIncludeBuildServerFeed());
+      vscode.window.showInformationMessage(
+        `${this.name} installed to ${locationString(selectedLocation)} successfully.`
+      );
+      return installedPackage;
+    } catch (error) {
+      Log.log(LOG_PREFIX, `Failed to install ${this.name}. ${error}`);
+      vscode.window.showErrorMessage(`Failed to install ${this.name}. ${error}`);
+      installedPackage = null;
+    }
+    return installedPackage;
+  }
+
+  async tryUpdate(
+    current: DotNetPackage,
+    target: DotNetPackage
+  ): Promise<{ updateResult: UpdateResult; package: DotNetPackage | null }> {
+    Log.log(
+      LOG_PREFIX,
+      `Already have ${current.name} with the correct version ${current.version.toString()}. ${current.location}.`
+    );
+    Log.log(LOG_PREFIX, `Offer to update ${current.name} if new version is available from nuget`);
+    if (current.version.compare(target.version) > 0) {
+      vscode.window.showInformationMessage(
+        `The ${this.name} ${current.version} on your box is newer version than the required version:${target.version}. Please check if there is a newer extension.`
+      );
+      return { updateResult: UpdateResult.currentPackageIsNewer, package: current };
+    }
+    if (current.version.compare(target.version) < 0 && current.version.isNewMajorOrMinorVersion(target.version)) {
+      this.context.workspaceState.update(this.neverShowAgainKey, false);
+    }
+    let neverShowAgain = this.context.workspaceState.get<boolean>(this.neverShowAgainKey);
+    if (neverShowAgain) {
+      return { updateResult: UpdateResult.declinedToUpdate, package: current };
+    }
+    return await this.handlePatchUpdate(current, target);
+  }
+
+  private async selectUserPreferredLocation(): Promise<PackageLocation | null> {
+    const moreInfoLink = new vscode.MarkdownString(`[More information](https://dotnet.microsoft.com/download)`);
+    moreInfoLink.isTrusted = true;
+    let response = await vscode.window.showInformationMessage(
+      `${this.name} ${this.version} is required. Where would you like to install it? ${moreInfoLink.value}`,
```

Comment:
THis is nit picky, but I'm not sure we should say "required". Maybe something like "This workspace contains .neo-express files but the Neo.Express tool package is not installed. Would you like to install it? (more info link)" with Local/Global/Dont ask again options

Reply:
text updated, thank you for the feedback